### PR TITLE
Http annotations

### PR DESCRIPTION
**Pull request by Raffo**
**WIP do not merge**

This is a possible implementation of http method tags. It is not intended to be merged as is, but it is mostly to show a possible implementation. I would appreciate your feedback @Vetcher :-) 

This replaces #48 

- This needs additional work for the client side, which I'll do after the feedback.

### CODE REVIEW

File: generator/template/http_server.go [hunk 1]
```diff
@@ -80,19 +101,25 @@ func (t *httpServerTemplate) Render() write_strategy.Renderer {
...
+		g.Id("mux").Op(":=").Qual(PackageGorillaMux, "NewRouter").Call()
 		for _, fn := range t.Info.Iface.Methods {
-			g.Id("handler").Dot("Handle").Call(
-				Lit("/"+util.ToURLSnakeCase(fn.Name)),
+			tags := util.FetchTags(fn.Docs, TagMark+httpMethodTag)
+			tag := ""
+			if len(tags) == 1 {
+				tag = strings.ToUpper(tags[0])
+			} else {
+				tag = "GET"
```
[the remark this answers is not in the document]
Comment (Raffo) [reply generator/template/http_server.go]: Probably better as it would allow us to pass a body. I'll change it.

**Comment by Raffo** [comment]
I changed the default method. 

There is clearly something else that needs to be changed: 

- The client side needs to be adapted. 
- In the case of GET request, it is a bug to try yo read the body. This means that we should change the code to make sure we don't do that otherwise we would just get a 500. 

Let me know what is you opinion on those changes.

**Comment by Vetcher** [comment]
For this GET cases we need some default url structure, e.g. `/{country}/{city}/{area}` for method 
```go
type Service interface {
    Houses(ctx context.Context, country, city, area string) (houses []House, err error)
}
```
and, perhaps, provide ability to setup custom url with variables (+ validation before generating).

**Comment by Raffo** [comment]
Your proposal sound nice, I would go for what you have proposed. 

What probably doesn't make sense to support is something like `/account/{name}/resource/{name}` a.k.a. subresources. I bet you want to keep it simple and avoid cases like that which are not directly mapped to go-kit's interface first approach.

**Comment by Vetcher** [comment]
I think we may add tag (something like `http-path` or `path`) for methods and some rules for parsing and validation.
Right now, Idk where we should add paths to urls on client side. In example [profilesvc](https://github.com/go-kit/kit/blob/master/examples/profilesvc/transport.go#L228), they add paths in encoders, but I have some doubts about this decision.

**Comment by Vetcher** [comment]
I'm going to add some commits to your pr.

**Comment by Raffo** [comment]
Sure, I left it open exactly for this. Again, let's have something minimal we can actually use and then iterate on that and not try to cover all use cases.

**Comment by Vetcher** [comment]
I duplicate your branch to https://github.com/devimteam/microgen/tree/http-annotations

**Comment by Raffo** [comment]
Cool, should I close this PR or shall we keep working on it? Can you summarize the status now? 

I would also love to add some tests for some of the methods.

**Comment by Vetcher** [comment]
You may close PR, but, please, create issues for each enhancement we mentioned here before. Right now, tests for templates is not best idea, because microgen still in intensive development, but I can add tests for utility functions, just list them.

File: generator/strings/strings.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package strings
+
+import "strings"
+
+// Fetch information from slice of comments (docs).
+// Returns appendix of first comment which has tag as prefix.
+func FetchMetaInfo(tag string, comments []string) string {
```
Comment (Raffo) [inline generator/strings/strings.go]: This function should be easily testable.

File: generator/template/http_converters.go [hunk 3]
```diff
@@ -290,3 +291,13 @@ func httpEncodeResponseName(f *types.Function) string {
 func httpDecodeResponseName(f *types.Function) string {
 	return "DecodeHTTP" + util.ToUpperFirst(f.Name) + "Response"
 }
+
+func encodeHttpRequestBody(fn *types.Function) *Statement {
```
Comment (Raffo) [inline generator/template/http_converters.go]: nitpick: `Http` should be `HTTP` as per Go guidelines.

**Comment by Raffo** [comment]
I'm not sure if it is worth closing this PR. I think the code is not done and I would love to keep contributing (and don't think you want to add me as contributor to your repo :-) ). In such a case I can keep adding some commits here and you can move them to your branch if you want... what do you think should I do as contributor to your project? 

For the tests, I can add some if you want, but I would rather focus now on finishing the main functionality of this feature.

**Comment by Raffo** [comment]
PR to which branch? Those changes are now merged into `devimteam:develop` so if I open it there will be no changes... or are you reverting them?

**Comment by Vetcher** [comment]
To develop, when you add new commits =)

**Comment by Vetcher** [comment]
Also, you may check 0.7.0 and tell your thoughts

**Comment by Raffo** [comment]
Sure, I will open another PR with the other changes and give feedback on 0.7.0 :-)
